### PR TITLE
Add new 'online' channel

### PR DESCRIPTION
**Pull request by M0rgan01**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | Dynamic calculation of the target version based on your PHP version. 
| Type?             | improvement
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | -
| Sponsor company   | -
| How to test?      | Make updates using the 2 channels

**BC breacks**

We will now have 2 channels for an update: `online` and `local`.

- `local` allows you to use a ZIP / XML pair, as before

- `online` allows dynamic calculation depending on your version of PHP

### CODE REVIEW

File: classes/Parameters/ConfigurationValidator.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+
+    /**
+     * @throws UpgradeException
+     */
+    private function validateChannel(string $channel): void
+    {
+        if ($channel !== Upgrader::CHANNEL_ARCHIVE && $channel !== Upgrader::CHANNEL_DYNAMIC) {
+            throw new UpgradeException('Unknown channel ' . $channel);
```
Comment (Quetzacoalt91) [inline classes/Parameters/ConfigurationValidator.php]:
```suggestion
            throw new UnexpectedValueException('Unknown channel ' . $channel);
```

File: classes/Services/PhpVersionResolverService.php [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        return [
+            new PrestashopRelease('1.7.0.0', 'stable', '7.1', '5.4'),
+            new PrestashopRelease('1.7.1.0', 'stable', '7.1', '5.4'),
+            new PrestashopRelease('1.7.2.0', 'stable', '7.1', '5.4'),
+            new PrestashopRelease('1.7.3.0', 'stable', '7.1', '5.4'),
+            new PrestashopRelease('1.7.4.0', 'stable', '7.1', '5.6'),
+            new PrestashopRelease('1.7.5.0', 'stable', '7.2', '5.6'),
+            new PrestashopRelease('1.7.6.0', 'stable', '7.2', '5.6'),
+            new PrestashopRelease('1.7.7.0', 'stable', '7.3', '5.6'),
+            new PrestashopRelease('1.7.8.0', 'stable', '7.4', '5.6'),
```
Comment (Quetzacoalt91) [inline classes/Services/PhpVersionResolverService.php]:
```suggestion
            '1.7.0' => new PrestashopRelease('1.7.0.0', 'stable', '7.1', '5.4'),
            '1.7.1' => new PrestashopRelease('1.7.1.0', 'stable', '7.1', '5.4'),
            // [...]
            '1.7.8' => new PrestashopRelease('1.7.8.0', 'stable', '7.4', '5.6'),
```

This would remove the foreach you have in the caller.

```php
            $targetMinorVersion = VersionUtils::splitPrestaShopVersion($targetVersion)['minor'];

            $requirements = $this->getPrestashop17Requirements();
            if (isset($requirements[$targetMinorVersion])) {
                $range = [
                    'php_min_version' => $requirements[$targetMinorVersion]->getPhpMinVersion(),
                    'php_max_version' => $requirements[$targetMinorVersion]->getPhpMaxVersion(),
                ];
            }
```

I prefer the previous version, the array allowed you to assign the compatibility range directly.

File: classes/Parameters/ConfigurationValidator.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/AFL-3.0 Academic Free License 3.0 (AFL-3.0)
+ */
+
+namespace PrestaShop\Module\AutoUpgrade\Parameters;
+
+use http\Exception\UnexpectedValueException;
```
Comment (Quetzacoalt91) [inline classes/Parameters/ConfigurationValidator.php]:
```suggestion
use UnexpectedValueException;
```

File: classes/Parameters/UpgradeConfiguration.php [hunk 4]
```diff
@@ -29,6 +29,7 @@
 
 use Configuration;
 use Doctrine\Common\Collections\ArrayCollection;
+use http\Exception\UnexpectedValueException;
```
Comment (Quetzacoalt91) [inline classes/Parameters/UpgradeConfiguration.php]:
```suggestion
use UnexpectedValueException;
```

File: classes/Task/Upgrade/UpgradeNow.php [hunk 5]
```diff
@@ -53,45 +53,26 @@ public function run(): int
...
-        if ($this->container->getUpgradeConfiguration()->get('channel') == 'private' && !$this->container->getUpgradeConfiguration()->get('private_allow_major')) {
-            $upgrader->checkPSVersion(false, ['private', 'minor']);
-        } else {
-            $upgrader->checkPSVersion();
-        }
-
         if ($upgrader->isLastVersion()) {
             $this->next = '';
-            $this->logger->info($this->translator->trans('You already have the %s version.', [$upgrader->version_name]));
+            $this->logger->info($this->translator->trans('You are using the most up to date version for your PHP version'));
```
Comment (Quetzacoalt91) [inline classes/Task/Upgrade/UpgradeNow.php]: Can we find something more generic? We may look for updates based on something else than the PHP version in the future.

File: classes/UpgradeContainer.php [hunk 6]
```diff
@@ -396,41 +398,23 @@ public function getUpgrader(): Upgrader
             define('_PS_ROOT_DIR_', $this->getProperty(self::PS_ROOT_PATH));
         }
 
-        $fileLoader = $this->getFileLoader();
+        $currentPrestashopVersion = $this->getProperty(self::PS_VERSION);
         // in order to not use Tools class
```
Comment (Quetzacoalt91) [inline classes/UpgradeContainer.php]:
```suggestion
```

File: classes/Upgrader.php [hunk 7]
```diff
@@ -85,138 +75,91 @@ public function __construct(string $version, FileLoader $fileLoader)
...
-            // no break
-        case 'rc':
-            $followed_channels[] = 'stable';
-            // no break
-        case 'minor':
-        case 'major':
-        case 'private':
-            $followed_channels[] = 'stable';
+        if ($this->channel !== self::CHANNEL_ONLINE) {
+            throw new LogicException('channel must be online to retrieve the version dynamically');
```
Comment (Quetzacoalt91) [inline classes/Upgrader.php]: Not sure I got this exception right.
* The message does not seem to be related to the method contents
* What's preventing the Upgrader in archive mode to eventually look for updates?

File: classes/Upgrader.php [hunk 8]
```diff
@@ -230,4 +173,9 @@ public function clearXmlMd5File(string $version): bool
 
         return true;
     }
+
+    public function getChannel(): string
```
Comment (Quetzacoalt91) [inline classes/Upgrader.php:175]:
```suggestion
    /**
     * @return self::CHANNEL_*
     */
    public function getChannel(): string
```

File: classes/Task/Upgrade/UpgradeNow.php (hunk 5 above)
Comment (Quetzacoalt91) [reply classes/Task/Upgrade/UpgradeNow.php]: "Your shop is currently running the latest compatible version. No updates are needed at this time."

File: classes/Upgrader.php (hunk 7 above)
Comment (M0rgan01) [reply classes/Upgrader.php]: For me it is important to understand that this function is reserved for online mode, so that it is not accidentally called on a local channel. It's a kind of safeguard. 

If you are in an upgrade process in archive mode, you should not use this function, you will not search for your archive but the online version.

In which case are you in archive mode and you also want to have the online mode release?

Comment (Quetzacoalt91) [reply classes/Upgrader.php]: I would imagine both channels details when refreshing the update version selection after selecting the archive.
You display the local channel as selected, but you still need to display what would be the updated version via the online channel.

Comment (M0rgan01) [reply classes/Upgrader.php]: The name of the method is enough for me indeed, and it is interesting to be able to compare the 2. I remove the condition

File: views/templates/block/upgradeButtonBlock.html.twig [hunk 9]
```diff
@@ -15,10 +13,9 @@
             <p>
                 {{ 'Your current PHP version'|trans({}) }}: <strong>{{ phpVersion }}</strong>
             </p>
-            <p>
-                {{ 'Latest official version for %s channel.'|trans([channel]) }}:
-                <strong>{{ latestChannelVersion }}</strong>
-            </p>
+            {% if not isLastVersion %}
+                <p>{{ 'Version chosen for update'|trans([channel]) }}: <strong>{{ destinationVersion }}</strong></p>
```
Comment (Quetzacoalt91) [inline views/templates/block/upgradeButtonBlock.html.twig]:
```suggestion
                <p>{{ 'Version available on channel %s:'|trans([channel]) }} <strong>{{ destinationVersion }}</strong></p>
```

File: classes/Upgrader.php [hunk 10]
```diff
@@ -27,56 +27,45 @@
...
-    const DEFAULT_CHANNEL = 'minor';
+    const DEFAULT_CHANNEL = self::CHANNEL_ONLINE;
     const DEFAULT_FILENAME = 'prestashop.zip';
 
-    /**
-     * @var bool contains true if last version is not installed
-     */
-    private $need_upgrade = false;
-
     /** @var string */
```
Comment (Quetzacoalt91) [inline classes/Upgrader.php]:
```suggestion
    /** @var self::CHANNEL_* */
```

**Review by Quetzacoalt91 — DISMISSED** [review]

**Review by Quetzacoalt91 — DISMISSED** [review]

**Review by Quetzacoalt91 — DISMISSED** [review]

**Review by ga-devfront — DISMISSED** [review]

**Comment by Quetzacoalt91** [comment]
While further tests are done, it appears selecting another channel does not refresh the page automatically anymore and I cannot switch to the `archive` channel.
An error can be found in the browser console:
image

On another version, the message appears properly but the selected option remains the same after refresh


[Capture vidéo du 2024-09-25 11-33-43.webm](https://github.com/user-attachments/assets/e0dd5b86-fb31-49c8-b675-b9374b30756c)

**Comment by Quetzacoalt91** [comment]
Tests done with this pull-requests:
PS : 8.1.0 - PHP : 8.1 - Upgrade to PS 8.1.7 Expected (via online channel): Successful ✅
PS : 1.7.3 - PHP : 7.1 - Upgrade to PS 1.7.8.11 Expected (via online channel): Successful ✅
PS : 1.7.5.2 - PHP : 7.2 - Upgrade to PS 8.2 Expected (via archive channel): Successful ✅
PS : 1.7.5.2 - PHP : 7.2 - Upgrade to PS 1.7.8.0 Expected (via archive channel): Successful ✅

Good to go!

**Review by Quetzacoalt91 — APPROVED** [review]
